### PR TITLE
Initialize native unwinder in 'PeCoffInterface' and call it in 'Step'

### PR DESCRIPTION
**Pull request by ronaldfw**
The native PE/COFF unwinder is essentially completely implemented. This
change adds one of the final missing pieces: Initializing the native
unwinder in 'PeCoffInterface' and call its 'Step' method as a fallback
to the .debug_frame 'Step' call. Some additional data needed to be
parsed from the PE/COFF object file, which required some changes to the
test class 'PeCoffFake' to make sure that 'PeCoffInterface::Init' still
succeeds. The native unwinder is only initialized in the 64-bit case,
as we do not have an implementation for 32-bit unwinding. The latter is
still possible using .debug_frame data if present.

Tested: Unit tests pass; ran sample application with unwinding.
Bug: http://b/194768602

### CODE REVIEW

File: third_party/libunwindstack/PeCoffInterface.cpp [hunk 1]
```diff
@@ -441,6 +448,59 @@ bool PeCoffInterfaceImpl<AddressType>::Init(int64_t* load_bias) {
   return true;
 }
 
+template <typename AddressType>
+bool PeCoffInterfaceImpl<AddressType>::MapFromRVAToFileOffset(uint64_t rva, uint64_t* file_offset) {
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/PeCoffInterface.cpp:452]: Yeah, that's also the Google code style. Will fix.

File: third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp [hunk 2]
```diff
@@ -401,42 +406,93 @@ class MockDwarfSection : public DwarfDebugFrame<AddressType> {
   MOCK_METHOD(bool, Step, (uint64_t, Regs*, Memory*, bool*, bool*), (override));
 };
 
+class MockPeCoffNativeUnwinder : public PeCoffNativeUnwinder {
+ public:
+  MockPeCoffNativeUnwinder() {}
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp:411]: Fixed.

File: third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp [hunk 3]
```diff
@@ -401,42 +406,93 @@ class MockDwarfSection : public DwarfDebugFrame<AddressType> {
...
+  virtual ~MockPeCoffNativeUnwinder() = default;
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp:412]: Fixed.

File: third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp [hunk 4]
```diff
@@ -401,42 +406,93 @@ class MockDwarfSection : public DwarfDebugFrame<AddressType> {
...
   ASSERT_TRUE(fake_coff.Init(&load_bias));
 
-  MockDwarfSection<typename TypeParam::AddressType>* mock_debug_frame_section =
-      new MockDwarfSection<typename TypeParam::AddressType>(nullptr);
-  fake_coff.SetFakeDebugFrameSection(mock_debug_frame_section);
-  EXPECT_CALL(*mock_debug_frame_section, Step(0x2000, nullptr, nullptr, nullptr, nullptr))
+  bool finished = false;
+  bool is_signal_frame = false;
+
+  std::unique_ptr<MockDwarfSection<typename TypeParam::AddressType>> mock_debug_frame_section =
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/tests/PeCoffInterfaceTest.cpp:426]: Yes, good suggestion. Fixed.

File: third_party/libunwindstack/CMakeLists.txt [hunk 5]
```diff
@@ -386,12 +388,12 @@ set_target_properties(libunwindstack_local PROPERTIES
 # Fuzz testing for parsers
 add_fuzzer(PeCoffInterfaceFuzzer tests/fuzz/PeCoffInterfaceFuzzer.cpp)
 target_include_directories(PeCoffInterfaceFuzzer PRIVATE ${CMAKE_CURRENT_SOURCE_DIR} include/)
-target_link_libraries(PeCoffInterfaceFuzzer PRIVATE libunwindstack)
+target_link_libraries(PeCoffInterfaceFuzzer PRIVATE CONAN_PKG::capstone libunwindstack)
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/CMakeLists.txt:391]: I removed it and it still builds. Maybe it was needed earlier for some reason, not sure.

File: third_party/libunwindstack/include/unwindstack/PeCoffInterface.h [hunk 6]
```diff
@@ -224,6 +227,8 @@ class PeCoffInterfaceImpl : public PeCoffInterface {
   bool GetSectionName(const std::string& parsed_section_name_string, std::string* result);
   bool InitSections();
   bool InitDebugFrameSection();
+  bool MapFromRVAToFileOffset(uint64_t rva, uint64_t* file_offset);
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/include/unwindstack/PeCoffInterface.h:230]: Yeah, libunwindstack seems to use this pattern throughout. That is, always a 'bool' return value to indicate success with results as in-out pointers. I don't have a strong preference here, but probably would prefer to keep it for consistency.

File: third_party/libunwindstack/tests/PeCoffFake.cpp [hunk 7]
```diff
@@ -36,20 +36,33 @@ void PeCoffFake<PeCoffInterfaceType>::Init() {
...
 
+  // .debug_frame section
   offset = SetDebugFrameEntryAtOffset(kDebugFrameSectionFileOffset, 0x2100);
   uint64_t debug_frame_vmsize = offset - kDebugFrameSectionFileOffset;
 
   // Invalid entry after the .debug_frame section. We want to validate that this entry does *not*
   // get parsed.
-  offset = SetDebugFrameEntryAtOffset(kDebugFrameSectionFileOffset + 0x200, 0x10000);
-  uint64_t debug_frame_filesize = offset - kDebugFrameSectionFileOffset;
+  uint64_t end_offset_debug_frame_section =
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/tests/PeCoffFake.cpp:49]: Fixed.

File: third_party/libunwindstack/tests/PeCoffFake.cpp [hunk 8]
```diff
@@ -36,20 +36,33 @@ void PeCoffFake<PeCoffInterfaceType>::Init() {
...
+      SetDebugFrameEntryAtOffset(kDebugFrameSectionFileOffset + 0x200, 0x10000);
+  uint64_t debug_frame_filesize = end_offset_debug_frame_section - kDebugFrameSectionFileOffset;
 
   CHECK(debug_frame_vmsize <= std::numeric_limits<uint32_t>::max());
   CHECK(debug_frame_filesize <= std::numeric_limits<uint32_t>::max());
-  SetSectionHeadersAtOffset(section_headers_offset, static_cast<uint32_t>(debug_frame_vmsize),
-                            static_cast<uint32_t>(debug_frame_filesize));
+  uint64_t end_offset_section_headers =
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/tests/PeCoffFake.cpp:55]: Fixed.

File: third_party/libunwindstack/tests/PeCoffFake.cpp [hunk 9]
```diff
@@ -229,12 +250,25 @@ uint64_t PeCoffFake<PeCoffInterfaceType>::SetOptionalHeaderAtOffset(uint64_t off
...
-    offset = SetData32(offset, 0);
-    offset = SetData32(offset, 0);
+  struct DataDirEntry {
+    uint32_t vmaddr;
+    uint32_t vmsize;
+  };
+
+  std::vector<DataDirEntry> kDataDirs{{0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}, {0, 0}};
+  constexpr int kCoffDataDirExceptionTableIndex = 3;
+  // The "exception table" and the ".pdata" section are the same thing.
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/tests/PeCoffFake.cpp:261]: The reason was that the PE format documentation calls this "Exception table", see https://docs.microsoft.com/en-us/windows/win32/debug/pe-format. But there's really no good other reason. I've changed to a consistent name, but added an explanation what the "official" name is.

File: third_party/libunwindstack/PeCoff.cpp [hunk 10]
```diff
@@ -154,7 +154,9 @@ bool PeCoff::Step(uint64_t rel_pc, Regs* regs, Memory* process_memory, bool* fin
                   bool* is_signal_frame) {
   // Lock during the step which can update information in the object.
   std::lock_guard<std::mutex> guard(lock_);
-  return interface_->Step(rel_pc, regs, process_memory, finished, is_signal_frame);
+
+  // TODO: Plumb pc_adjustment.
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/PeCoff.cpp:158]: No, this has to be still done in some other places, including the public interface of the Object class ('Step' method).

I could leave the pc_adjustment part out here, as it's just called with 0 anyway. That would be a bit cleaner, but wouldn't make a difference once the next changes are in.

**Comment by ronaldfw** [comment]
Thanks for the review so far @dpallotti, PTAL

**Review by dpallotti — APPROVED** [review]

File: third_party/libunwindstack/include/unwindstack/PeCoffInterface.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (dpallotti) [reply third_party/libunwindstack/include/unwindstack/PeCoffInterface.h]: Consistency sounds good.

File: third_party/libunwindstack/PeCoffInterface.cpp [hunk 11]
```diff
@@ -441,6 +448,59 @@ bool PeCoffInterfaceImpl<AddressType>::Init(int64_t* load_bias) {
...
+      return true;
+    }
+  }
+  last_error_.code = ERROR_INVALID_COFF;
+  return false;
+}
+
+template <>
+bool PeCoffInterfaceImpl<uint64_t>::InitNativeUnwinder() {
+  constexpr int kCoffDataDirExceptionTableIndex = 3;
```
[the remark this answers is not in the document]
Comment (ronaldfw) [reply third_party/libunwindstack/PeCoffInterface.cpp:465]: I think this should be "exception table" here as this is about parsing the header. For consistency, I've changed it to exception table in the PeCoffFake class as well.

**Pull request merged**